tests/communications/test_notification_preferences.py

```diff
@@ -0,0 +1,156 @@
+            code='test_type_1',
+            name='Test Type 1',
+            description='Test Description 1',
+            default_frequency='immediate'
+        )
+        self.notification_type2 = NotificationType.objects.create(
+            code='test_type_2',
+            name='Test Type 2',
+            description='Test Description 2',
+            default_frequency='daily_digest'
+        )
+
+    def test_get_notification_types(self):
+        """Test retrieving notification types."""
+        url = reverse('communications:notification-type-list')
+        response = self.client.get(url)
+        
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIsInstance(response.data, list)
+        codes = [item['code'] for item in response.data]
+        self.assertIn(self.notification_type1.code, codes)
+        self.assertIn(self.notification_type2.code, codes)
+
+    def test_get_user_preferences(self):
+        """Test retrieving user notification preferences."""
+        url = reverse('communications:user-notification-preference-list')
+        response = self.client.get(url)
+        
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertIsInstance(response.data, list)
+        self.assertTrue(len(response.data) > 0)
+        
+        user_prefs = response.data[0]
+        self.assertTrue('user_id' in user_prefs)
+        self.assertEqual(user_prefs['user_id'], self.user.pk)
+        self.assertTrue('enable_in_app' in user_prefs)
+        self.assertTrue('enable_email' in user_prefs)
+        self.assertTrue('enable_push' in user_prefs)
+        
+        self.assertIn('type_preferences', user_prefs)
+        self.assertIsInstance(user_prefs['type_preferences'], list)
+
+    def test_update_user_preferences(self):
+        """Test updating user notification preferences."""
+        pref = UserNotificationPreference.objects.get(user=self.user)
+        url = reverse('communications:user-notification-preference-detail', 
+                     kwargs={'pk': self.user.pk})
+        
+        data = {
+            'enable_in_app': True,
+            'enable_email': False,
+            'enable_push': True
+        }
+        response = self.client.patch(url, data, format='json')
+        
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertTrue(response.data['enable_in_app'])
+        self.assertFalse(response.data['enable_email'])
+        self.assertTrue(response.data['enable_push'])
+        
+        pref.refresh_from_db()
+        self.assertTrue(pref.enable_in_app)
+        self.assertFalse(pref.enable_email)
+        self.assertTrue(pref.enable_push)
+
+    def test_update_notification_type_preference(self):
+        """Test updating a specific notification type preference."""
+        pref = UserNotificationPreference.objects.get(user=self.user)
+        type_pref = pref.type_preferences.first()
```

Comment:
pref.type_preferences.first() assumes at least one type preference exists. If setup or signal creation fails, this will raise an AttributeError. Add assertions to confirm type_preferences.count() > 0 before accessing first().